### DISCUSSION TITLE
container & sessions

### DISCUSSION THREAD

**Discussion by mukral** [zauberzeug/nicegui#215] [Q&A] [answered]
Hi, no issue regarding nicegui more about a lack of lack programming skills...

I can't bring together programming multiuser sessions and page designing with containers and cards.
My goal is to show a scene which can be changed by each page visitor individually and the result is shown in the browser.
Each request for it's own works fine: when container/cards are at "top level" (without sesisions) it's possible to clear and the content of the scene and it starts on a clean scene. But each user changes the scene. When sessions are implemented I dont find a way to clear the "scene-container" so the scene is displayed at each call. It's totally clear to me why both happens - it's about the enclosed def-routines and with-statements. But I can't find a way how to solve it... maybe someone has a short hint?
```
from nicegui import ui
from starlette.middleware.sessions import SessionMiddleware
from starlette.requests import Request
class Demo:
    def __init__(self):
        self.number = 1
@ui.page('/startingPage')
def startingPage(request: Request):
    demo = Demo()
    with ui.row():
        with ui.card() as card1:
            select1 = ui.select([1, 2, 3], value=1).bind_value(demo, 'number')
        with ui.card() as card2:
            ui.button('go', on_click=lambda: buildScene(select1.value))
        with ui.card() as card3:
            ui.label('card3')
    print (request.values)
def buildScene(select1):
    colorSphere='#ffffff'
    if select1==1:
        colorSphere='#ff0000'
    elif select1==2:
        colorSphere='#00ff00'
    elif select1==3:
        colorSphere='#0000ff'
    card3.clear()
    with card3:
        with ui.scene() as scene:
            scene.sphere(2).material(colorSphere)
#    with ui.scene() as scene:
#        scene.sphere(2).material(colorSphere)
ui.link('link',startingPage)
ui.run()
```
I know it's more a newbie programming question than a nicegui topic (maybe an good exanple for other newbies).
So feel free to close the issue...

Thank you
Markus

**Comment by rodja** [comment, marked as answer]
Thanks for reaching out. I converted the issue into a discussion to better reflect the purpose.
I'm not entirely sure what you want to accomplish. Do you want one 3d scene which can be manipulated by multiple users (eg. browser windows)? Or do you want each scene to be independent of all the others? To make the second happen, try indenting the `buildScene` function to be a sub-function of `startingPage`. This way each client has it's own `buildScene` function which then can affect the cards created for their request.

**Reply by mukral** [reply]
Thank you @rodja.
It was the latter one - each scene to be independent of all other.
Your solution works great. I'm such a newbie, I didn't even think of sub-functions.
This is the working code:
```
from nicegui import ui
from starlette.middleware.sessions import SessionMiddleware
from starlette.requests import Request
class Demo:
    def __init__(self):
        self.number = 1
@ui.page('/startingPage')
def startingPage(request: Request):
    def buildScene(select1):
        colorSphere='#ffffff'
        if select1==1:
            colorSphere='#ff0000'
        elif select1==2:
            colorSphere='#00ff00'
        elif select1==3:
            colorSphere='#0000ff'
        card3.clear()
        with card3:
            with ui.scene() as scene:
                scene.sphere(2).material(colorSphere)
    demo = Demo()
    with ui.row():
        with ui.card() as card1:
            select1 = ui.select([1, 2, 3], value=1).bind_value(demo, 'number')
        with ui.card() as card2:
            ui.button('go', on_click=lambda: buildScene(select1.value))
        with ui.card() as card3:
            ui.label('card3')
    print (request.values)
ui.link('link',startingPage)
ui.run()
```
